src/robin_dict.jl

```diff
@@ -0,0 +1,593 @@
+             getindex, getkey, haskey, iterate, @propagate_inbounds,
+             pop!, delete!, get, get!, isbitstype, in, hashindex,
+             isiterable, dict_with_eltype, KeySet, Callable, _tablesz
+
+# the load factor arter which the dictionary `rehash` happens
+const ROBIN_DICT_LOAD_FACTOR = 0.70
+
+"""
+    RobinDict([itr])
+
+`RobinDict{K,V}()` constructs a hash table with keys of type `K` and values of type `V`.
+Keys are compared with [`isequal`](@ref) and hashed with [`hash`](@ref).
+Given a single iterable argument, constructs a [`RobinDict`](@ref) whose key-value pairs
+are taken from 2-tuples `(key,value)` generated by the argument.
+
+# Examples
+```jldoctest
+julia> RobinDict([("A", 1), ("B", 2)])
+RobinDict{String,Int64} with 2 entries:
+  "B" => 2
+  "A" => 1
+```
+
+Alternatively, a sequence of pair arguments may be passed.
+
+```jldoctest
+julia> RobinDict("A"=>1, "B"=>2)
+RobinDict{String,Int64} with 2 entries:
+  "B" => 2
+  "A" => 1
+```
+"""
+mutable struct RobinDict{K,V} <: AbstractDict{K,V}
+    #there is no need to maintain an table_size as an additional variable
+    # dibs::Array{Int8,1} # distance to initial bucket - critical for implementation
+    hashes::Vector{UInt32}
+    keys::Array{K,1}
+    vals::Array{V,1}
+    count::Int
+    totalcost::Int
+    maxprobe::Int   # length of longest probe
+    idxfloor::Int
+
+    function RobinDict{K, V}() where {K, V}
+        n = 16 # default size of an empty Dict in Julia
+        new(zeros(UInt32, n), Vector{K}(undef, n), Vector{V}(undef, n), 0, 0, 0, 0)
+    end
+
+    function RobinDict{K, V}(d::RobinDict{K, V}) where {K, V}
+        new(copy(d.hashes), copy(d.keys), copy(d.vals), d.count, d.totalcost, d.maxprobe, d.idxfloor)
+    end
+
+    function RobinDict{K, V}(keys, vals, hashes, count, totalcost, maxprobe, idxfloor) where {K, V}
+        new(hashes, keys, vals, count, totalcost, maxprobe, idxfloor)
+    end
+end
+
+function RobinDict{K,V}(kv) where V where K
+    h = RobinDict{K,V}()
+    for (k,v) in kv
+        h[k] = v
+    end
+    return h
+end
+RobinDict{K,V}(p::Pair) where {K,V} = setindex!(RobinDict{K,V}(), p.second, p.first)
+function RobinDict{K,V}(ps::Pair...) where V where K
+    h = RobinDict{K,V}()
+    sizehint!(h, length(ps))
+    for p in ps
+        h[p.first] = p.second
+    end
+    return h
+end
+
+RobinDict() = RobinDict{Any,Any}()
+RobinDict(kv::Tuple{}) = RobinDict()
+copy(d::RobinDict) = RobinDict(d)
+
+RobinDict(ps::Pair{K,V}...) where {K,V} = RobinDict{K,V}(ps)
+RobinDict(ps::Pair...)                  = RobinDict(ps)
+
+function RobinDict(d::AbstractDict{K, V}) where {K, V}
+    h = RobinDict{K, V}()
+    for (k, v) in d
+        h[k] = v
+    end
+    h
+end
+
+function RobinDict(kv)
+    try
+        d = dict_with_eltype((K, V) -> RobinDict{K, V}, kv, eltype(kv))
+        if isa(d, Dict)
+            return RobinDict(d)
+        else
+            return d
+        end
```

Comment:
there were few tests for L93 that were returning `Dict` instead of `RobinDict`, so the whole code segment checks for the return type of `dict_with_eltype` and acts accordingly

Reply:
Are you able to reproduce those? It sounds like a bug in `dict_with_eltype`, unless you perhaps just had old state confusing the test?

Reply:
Test at L132, 140 and 148 takes a hit, if you just return what `dict_with_eltype` returns.
```julia
typeof: Test Failed at C:\Users\ww\Desktop\Julia\DataStructures.jl\test\test_robin_dict.jl:132
  Expression: typeof(d) == typeof(d2) == typeof(d3) == RobinDict{Int, Any}
   Evaluated: RobinDict{Int64,Any} == Dict{Int64,Any} == Dict{Int64,Any} == RobinDict{Int64,Any}
Stacktrace:
 [1] top-level scope at C:\Users\ww\Desktop\Julia\DataStructures.jl\test\test_robin_dict.jl:132
 [2] top-level scope at C:\cygwin\home\Administrator\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v1.1\Test\src\Test.jl:1083
 [3] top-level scope at C:\Users\ww\Desktop\Julia\DataStructures.jl\test\test_robin_dict.jl:118
typeof: Test Failed at C:\Users\ww\Desktop\Julia\DataStructures.jl\test\test_robin_dict.jl:140
  Expression: typeof(d) == typeof(d2) == typeof(d3) == RobinDict{Any, Int}
   Evaluated: RobinDict{Any,Int64} == Dict{Any,Int64} == Dict{Any,Int64} == RobinDict{Any,Int64}
Stacktrace:
 [1] top-level scope at C:\Users\ww\Desktop\Julia\DataStructures.jl\test\test_robin_dict.jl:140
 [2] top-level scope at C:\cygwin\home\Administrator\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v1.1\Test\src\Test.jl:1083
 [3] top-level scope at C:\Users\ww\Desktop\Julia\DataStructures.jl\test\test_robin_dict.jl:118
typeof: Test Failed at C:\Users\ww\Desktop\Julia\DataStructures.jl\test\test_robin_dict.jl:148
  Expression: typeof(d) == typeof(d2) == typeof(d3) == RobinDict{Any, Any}
   Evaluated: RobinDict{Any,Any} == Dict{Any,Any} == Dict{Any,Any} == RobinDict{Any,Any}
Stacktrace:
 [1] top-level scope at C:\Users\ww\Desktop\Julia\DataStructures.jl\test\test_robin_dict.jl:148
 [2] top-level scope at C:\cygwin\home\Administrator\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v1.1\Test\src\Test.jl:1083
 [3] top-level scope at C:\Users\ww\Desktop\Julia\DataStructures.jl\test\test_robin_dict.jl:118
Test Summary: | Pass  Fail  Total
typeof        |   13     3     16
ERROR: LoadError: Some tests did not pass: 13 passed, 3 failed, 0 errored, 0 broken.
```

Reply:
Probably some bug in the code of `dict_with_eltype`

Reply:
Looks like you're missing:
```julia
empty(d::RobinDict, ::Type{K}, ::Type{V}) where {K, V} = RobinDict{K, V}()
```